File: src/utils/indicators.js
// Technical indicators computed client-side from a close-price series.
// Pure functions, no deps. Used by StockDetail + the Buy/Hold/Sell badge.

export function sma(values, period) {
  if (!Array.isArray(values) || values.length < period || period <= 0) return null;
  let sum = 0;
  for (let i = values.length - period; i < values.length; i++) sum += values[i];
  return sum / period;
}

export function ema(values, period) {
  if (!Array.isArray(values) || values.length < period || period <= 0) return null;
  const k = 2 / (period + 1);
  // Seed with SMA of the first `period` values, then walk forward.
  let prev = values.slice(0, period).reduce((a, b) => a + b, 0) / period;
  for (let i = period; i < values.length; i++) prev = values[i] * k + prev * (1 - k);
  return prev;
}

// Full EMA series (needed for MACD signal line).
function emaSeries(values, period) {
  if (!Array.isArray(values) || values.length < period || period <= 0) return [];
  const k = 2 / (period + 1);
  const out = [];
  let prev = values.slice(0, period).reduce((a, b) => a + b, 0) / period;
  out.push(prev);
  for (let i = period; i < values.length; i++) {
    prev = values[i] * k + prev * (1 - k);
    out.push(prev);
  }
  return out;
}

// Wilder's RSI. Returns 0-100 or null.
export function rsi(values, period = 14) {
  if (!Array.isArray(values) || values.length < period + 1) return null;
  let gain = 0;
  let loss = 0;
  for (let i = 1; i <= period; i++) {
    const diff = values[i] - values[i - 1];
    if (diff >= 0) gain += diff; else loss -= diff;
  }
  let avgGain = gain / period;
  let avgLoss = loss / period;
  for (let i = period + 1; i < values.length; i++) {
    const diff = values[i] - values[i - 1];
    const up = diff > 0 ? diff : 0;
    const down = diff < 0 ? -diff : 0;
    avgGain = (avgGain * (period - 1) + up) / period;
    avgLoss = (avgLoss * (period - 1) + down) / period;
  }
  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

// MACD line, signal line, histogram (12/26/9 default).
export function macd(values, fast = 12, slow = 26, signal = 9) {
  if (!Array.isArray(values) || values.length < slow + signal) return null;
  const fastSeries = emaSeries(values, fast);
  const slowSeries = emaSeries(values, slow);
  // Align tails (slow series is shorter by slow-fast).
  const offset = fastSeries.length - slowSeries.length;
  const macdLine = slowSeries.map((s, i) => fastSeries[i + offset] - s);
  const signalLine = emaSeries(macdLine, signal);
  const macdVal = macdLine[macdLine.length - 1];
  const signalVal = signalLine[signalLine.length - 1];
  return { macd: macdVal, signal: signalVal, histogram: macdVal - signalVal };
}

// Bollinger Bands (20-period, 2 stddev default).
export function bollinger(values, period = 20, mult = 2) {
  if (!Array.isArray(values) || values.length < period) return null;
  const slice = values.slice(values.length - period);
  const mean = slice.reduce((a, b) => a + b, 0) / period;
  const variance = slice.reduce((a, b) => a + (b - mean) ** 2, 0) / period;
  const sd = Math.sqrt(variance);
  return { middle: mean, upper: mean + mult * sd, lower: mean - mult * sd };
}

// Composite Buy/Hold/Sell signal from RSI + MACD + 50/200 MA trend.
// Returns { label: 'Buy'|'Hold'|'Sell', score, reasons: [] }.
export function signal(values) {
  if (!Array.isArray(values) || values.length < 35) return null;
  let score = 0;
  const reasons = [];

  const r = rsi(values);
  if (r != null) {
    // Momentum reading: above 55 = strength, below 45 = weakness. Flag the
    // extremes as a caution in the reasons text without flipping the score.
    if (r > 55) { score += 1; reasons.push(r > 70 ? `RSI ${r.toFixed(0)} (strong, overbought)` : `RSI ${r.toFixed(0)} strong`); }
    else if (r < 45) { score -= 1; reasons.push(r < 30 ? `RSI ${r.toFixed(0)} (weak, oversold)` : `RSI ${r.toFixed(0)} weak`); }
  }

  const m = macd(values);
  if (m) {
    const eps = 1e-6;
    if (m.histogram > eps) { score += 1; reasons.push('MACD bullish'); }
    else if (m.histogram < -eps) { score -= 1; reasons.push('MACD bearish'); }
  }

  const fast = sma(values, 50);
  const slow = values.length >= 200 ? sma(values, 200) : sma(values, Math.min(100, values.length - 1));
  if (fast != null && slow != null) {
    if (fast > slow) { score += 1; reasons.push('Uptrend (MA)'); }
    else if (fast < slow) { score -= 1; reasons.push('Downtrend (MA)'); }
  }

  const label = score >= 2 ? 'Buy' : score <= -2 ? 'Sell' : 'Hold';
  return { label, score, reasons };
}

File: Shared/Indicators.swift
import Foundation

enum ChartType: String, CaseIterable {
    case line = "Line"
    case area = "Area"
    case candles = "Candles"
    case heikinAshi = "Heikin Ashi"
}

enum Indicators {
    struct DataPoint {
        let date: Date
        let value: Double
    }

    struct OHLCPoint: Identifiable {
        let date: Date
        let open: Double
        let high: Double
        let low: Double
        let close: Double
        var id: Date { date }

        var isBullish: Bool { close >= open }
    }

    static func heikinAshi(from points: [OHLCPoint]) -> [OHLCPoint] {
        guard !points.isEmpty else { return [] }
        var result: [OHLCPoint] = []
        result.reserveCapacity(points.count)
        for (i, p) in points.enumerated() {
            let haClose = (p.open + p.high + p.low + p.close) / 4.0
            let haOpen: Double
            if i == 0 {
                haOpen = (p.open + p.close) / 2.0
            } else {
                haOpen = (result[i - 1].open + result[i - 1].close) / 2.0
            }
            let haHigh = max(p.high, haOpen, haClose)
            let haLow = min(p.low, haOpen, haClose)
            result.append(OHLCPoint(date: p.date, open: haOpen, high: haHigh, low: haLow, close: haClose))
        }
        return result
    }

    static func sma(prices: [(Date, Double)], period: Int) -> [DataPoint] {
        guard prices.count >= period, period > 0 else { return [] }
        var result: [DataPoint] = []
        result.reserveCapacity(prices.count - period + 1)
        var sum = prices[0..<period].map(\.1).reduce(0, +)
        result.append(DataPoint(date: prices[period - 1].0, value: sum / Double(period)))
        for i in period..<prices.count {
            sum += prices[i].1 - prices[i - period].1
            result.append(DataPoint(date: prices[i].0, value: sum / Double(period)))
        }
        return result
    }

    static func ema(prices: [(Date, Double)], period: Int) -> [DataPoint] {
        guard prices.count >= period else { return [] }
        let k = 2.0 / Double(period + 1)
        // Seed with SMA of first `period` values
        let seed = prices[0..<period].map(\.1).reduce(0, +) / Double(period)
        var result: [DataPoint] = [DataPoint(date: prices[period - 1].0, value: seed)]
        for i in period..<prices.count {
            let prev = result.last!.value
            let ema = prices[i].1 * k + prev * (1 - k)
            result.append(DataPoint(date: prices[i].0, value: ema))
        }
        return result
    }

    // Buy/Hold/Sell signal from RSI + MACD + MA trend. Mirrors
    // src/utils/indicators.js so web and native agree.
    enum TradeSignal: String { case buy = "Buy", hold = "Hold", sell = "Sell" }

    private static func emaSeries(_ values: [Double], _ period: Int) -> [Double] {
        guard values.count >= period, period > 0 else { return [] }
        let k = 2.0 / Double(period + 1)
        var prev = values[0..<period].reduce(0, +) / Double(period)
        var out = [prev]
        for i in period..<values.count {
            prev = values[i] * k + prev * (1 - k)
            out.append(prev)
        }
        return out
    }

    private static func smaValue(_ values: [Double], _ period: Int) -> Double? {
        guard values.count >= period, period > 0 else { return nil }
        return values[(values.count - period)...].reduce(0, +) / Double(period)
    }

    static func rsi(_ values: [Double], period: Int = 14) -> Double? {
        guard values.count >= period + 1 else { return nil }
        var gain = 0.0, loss = 0.0
        for i in 1...period {
            let d = values[i] - values[i - 1]
            if d >= 0 { gain += d } else { loss -= d }
        }
        var avgGain = gain / Double(period), avgLoss = loss / Double(period)
        if values.count > period + 1 {
            for i in (period + 1)..<values.count {
                let d = values[i] - values[i - 1]
                avgGain = (avgGain * Double(period - 1) + max(d, 0)) / Double(period)
                avgLoss = (avgLoss * Double(period - 1) + max(-d, 0)) / Double(period)
            }
        }
        if avgLoss == 0 { return 100 }
        return 100 - 100 / (1 + avgGain / avgLoss)
    }

    static func macdHistogram(_ values: [Double], fast: Int = 12, slow: Int = 26, signalP: Int = 9) -> Double? {
        guard values.count >= slow + signalP else { return nil }
        let fastS = emaSeries(values, fast), slowS = emaSeries(values, slow)
        guard !fastS.isEmpty, !slowS.isEmpty else { return nil }
        let offset = fastS.count - slowS.count
        let macdLine = slowS.enumerated().map { fastS[$0.offset + offset] - $0.element }
        let signalLine = emaSeries(macdLine, signalP)
        guard let m = macdLine.last, let s = signalLine.last else { return nil }
        return m - s
    }

    static func signal(closes: [Double]) -> TradeSignal? {
        guard closes.count >= 35 else { return nil }
        var score = 0
        if let r = rsi(closes) {
            if r > 55 { score += 1 } else if r < 45 { score -= 1 }
        }
        if let h = macdHistogram(closes) {
            if h > 1e-6 { score += 1 } else if h < -1e-6 { score -= 1 }
        }
        let fast = smaValue(closes, 50)
        let slow = closes.count >= 200 ? smaValue(closes, 200) : smaValue(closes, min(100, closes.count - 1))
        if let f = fast, let s = slow {
            if f > s { score += 1 } else if f < s { score -= 1 }
        }
        return score >= 2 ? .buy : (score <= -2 ? .sell : .hold)
    }
}
